File: star-syncplayer/src/main/java/top/tonydon/util/URIUtils.java
package top.tonydon.util;

public class URIUtils {

    public static String encoder(String uri) {
        char[] charArray = uri.toCharArray();
        StringBuilder sb = new StringBuilder();
        for (char ch : charArray) {
            if (ch == ' ') {
                sb.append("%20");
            } else if (ch == '?') {
                sb.append("%3F");
            } else if (ch == '%') {
                sb.append("%25");
            } else if (ch == '#') {
                sb.append("%23");
            } else if (ch == '&') {
                sb.append("%26");
            } else if (ch == '=') {
                sb.append("%3D");
            } else if (ch == '{') {
                sb.append("%7B");
            } else if (ch == '}') {
                sb.append("%7D");
            } else if (ch == '"') {
                sb.append("%22");
            } else {
                sb.append(ch);
            }
        }
        return sb.toString();
    }
}

File: star-syncplayer-client/src/main/java/top/tonydon/syncplayer/util/URIUtils.kt
package top.tonydon.syncplayer.util

object URIUtils {
    fun encoder(uri: String): String {
        val charArray = uri.toCharArray()
        val sb = StringBuilder()
        for (ch in charArray) {
            when (ch) {
                ' ' -> sb.append("%20")
                '?' -> sb.append("%3F")
                '%' -> sb.append("%25")
                '#' -> sb.append("%23")
                '&' -> sb.append("%26")
                '=' -> sb.append("%3D")
                '{' -> sb.append("%7B")
                '}' -> sb.append("%7D")
                '"' -> sb.append("%22")
                else -> sb.append(ch)
            }
        }
        return sb.toString()
    }
}
